propkatraj/propkatraj.py

```diff
@@ -2,22 +2,239 @@
+    in order to alter the starting frame, final frame and frame sampling
+    frequency. Please note that frame numbers are zero formatted, and the
+    ``stop`` keyword is exclusive, as is the default in
+    :class:`MDAnalysis.AnalysisBase` objects. For example, in order to sample
+    every two frames from the first 10 frames::
+
+      pkatraj.run(start=0, stop=10, step=2)
+
+    You can also set the ``verbose`` keyword in order to have a nice progress
+    bar of the analysis progress::
+
+      pkatraj.run(verbose=True)
+
+    Once completed, the resulting timeseries of predicted pKas per residue is
+    made available as a :class:`pandas.DataFrame` under :attr`PropkaTraj`.pkas.
+    For example if one wanted to get a summary of the statistics of the
+    timeseries::
+
+      pkatraj.pkas.describe()
+
+    If one wanted to plot per residue boxplots of the timeseries data and save
+    it as a file `pKa-plot.png` (note: this requires the ``matplotlib`` and
+    ``seaborn`` packages which are not installed by default with
+    ``propkatraj``)::
+
+      import matplotlib.pyplot as plt
+      import seaborn as sns
+
+      fig = plt.figure(figsize=(28, 8))
+      ax = fig.add_subplot(1, 1, 1)
+      sns.boxplot(data=pkatraj.pkas, ax=ax)
+      ax.set_xlabel('residue number')
+      ax.set_ylabel(r'p$K_a$')
+      fig.savefig('pKa-plot.png')
+
+
+    .. versionadded:: 1.0.3
+    """
+    def __init__(self, atomgroup, select='protein', skip_failure=False,
+                 **kwargs):
+        if select is not None:
+            self.ag = atomgroup.select_atoms(select).atoms
+        else:
+            self.ag = atomgroup.atoms
+
+        # Issue #23 (keep until the PDBWriter is fixed)
+        if len(self.ag.select_atoms('not protein')) > 0:
+            wmsg = ("Non protein atoms passed to propka 3.1.\n MDAnalysis' "
+                    "PDBWriter does not currently write non-standard residues "
+                    "correctly as HETATM records and this may lead to "
+                    "incorrect pKa predictions.\n"
+                    "See https://github.com/Becksteinlab/propkatraj/issues/24 "
+                    " for more details")
+            warnings.warn(wmsg)
+
+        # TODO: probably needs to be a temporary directory instead
+        self.tmpfile = os.path.join(os.path.curdir, 'current.pdb')
+        self.skip_failure = skip_failure
+        super(PropkaTraj, self).__init__(self.ag.universe.trajectory, **kwargs)
+
+    def _prepare(self):
+        self._pkas = []
+        self.num_failed_frames = 0
+        self.failed_frames_log = []
+        self.failed_times = []
+        self._columns = None
+
+    def _single_frame(self):
+        pstream = mda.lib.util.NamedStream(StringIO(), self.tmpfile)
+        self.ag.write(pstream)
+        # reset stream for reading
+        pstream.reset()
+
+        try:
+            # TODO: it would be nice to allow for other options, maybe for 3.2?
```

Comment:
I guess so, I think we didn't need any others so we kept it simple.

Reply:
The only thing I could think could be useful here is the `-k` flag, but honestly I've never really been convinced keeping protons is ever worth it. I'll leave the comment in for now and we can revisit it post switch to 3.2?